### ISSUE TITLE
ops: modernize terraform setup

### ISSUE THREAD

**Issue by martinklepsch** [cljdoc/cljdoc#796]
(in part via #795)

When initially setting up cljdoc infrastructure I picked Terraform as a tool to declaratively describe our infrastructure and manage its state. Now, after years of it running without many modifications, the terraform DSL and packages have changed and it seems to be in a somewhat broken state.

Running `cd ops/infrastructure; terraform init` will spew various warnings about upgrading to terraform `0.13` (see below for a full print of those). At some point I attempted upgrading our Terraform setup to 0.13 but many of the issues I ran into felt a bit cryptic and weren't easily solvable by just searching the web a bit. 

I might give this another go soon but if you want to try @lread, I think making changes to successfully run `terraform init` would be a helpful first step. 

### Current output of `terraform init`
```
 terraform init

Initializing the backend...
Initializing modules...
╷
│ Warning: Quoted references are deprecated
│
│   on cljdoc.tf line 46, in resource "aws_route53_zone" "cljdoc_org_zone":
│   46:   provider = "aws.prod"
│
│ In this context, references are expected literally rather than in quotes. Terraform 0.11 and earlier required quotes, but quoted references are now
│ deprecated and will be removed in a future version of Terraform. Remove the quotes surrounding this reference to silence this warning.
│
│ (and 14 more similar warnings elsewhere)
╵

╷
│ Error: Invalid legacy provider address
│
│ This configuration or its associated state refers to the unqualified provider "aws".
│
│ You must complete the Terraform 0.13 upgrade process before upgrading to later versions.
╵

╷
│ Error: Invalid legacy provider address
│
│ This configuration or its associated state refers to the unqualified provider "digitalocean".
│
│ You must complete the Terraform 0.13 upgrade process before upgrading to later versions.
╵

╷
│ Error: Invalid legacy provider address
│
│ This configuration or its associated state refers to the unqualified provider "random".
│
│ You must complete the Terraform 0.13 upgrade process before upgrading to later versions.
```

**Comment by lread** [comment]
Ya, happy to take a peek sometime.

**Comment by lread** [comment]
I'm not sure what version of terraform we were using, but last commits on terraform related stuff is late 2018. So probably either 0.11.x or 0.12.x.

I found [tfenv](https://github.com/tfutils/tfenv) which allows me to install and switch between multiple versions of terraform.

I'll install the latest available terraform for each 0.minor version and upgrade slowly. 

## v0.11.15
A `terraform init` with the last 0.11 release (0.11.15) works.

```
Initializing modules...
- module.releases_bucket
  Getting source "./public_bucket"
- module.backups_bucket
  Getting source "./public_bucket"
- module.main_server
  Getting source "./server_instance"

Initializing provider plugins...
- Checking for available provider plugins on https://releases.hashicorp.com...
- Downloading plugin for provider "aws" (2.70.4)...
- Downloading plugin for provider "digitalocean" (2.29.0)...

The following providers do not have any version constraints in configuration,
so the latest version was installed.

To prevent automatic upgrades to new major versions that may contain breaking
changes, it is recommended to add version = "..." constraints to the
corresponding provider blocks in configuration, with the constraint strings
suggested below.

* provider.aws: version = "~> 2.70"
* provider.digitalocean: version = "~> 2.29"

Terraform has been successfully initialized!

You may now begin working with Terraform. Try running "terraform plan" to see
any changes that are required for your infrastructure. All Terraform commands
should now work.

If you ever set or change modules or backend configuration for Terraform,
rerun this command to reinitialize your working directory. If you forget, other
commands will detect it and remind you to do so if necessary.
```

## v0.12.31
```
Initializing modules...

Initializing the backend...

Initializing provider plugins...

The following providers do not have any version constraints in configuration,
so the latest version was installed.

To prevent automatic upgrades to new major versions that may contain breaking
changes, it is recommended to add version = "..." constraints to the
corresponding provider blocks in configuration, with the constraint strings
suggested below.

* provider.aws: version = "~> 2.70"
* provider.digitalocean: version = "~> 2.29"


Warning: Interpolation-only expressions are deprecated

  on cljdoc.tf line 6, in provider "aws":
   6:   region     = "${var.aws_region}"

Terraform 0.11 and earlier required all non-constant expressions to be
provided via interpolation syntax, but this pattern is now deprecated. To
silence this warning, remove the "${ sequence from the start and the }"
sequence from the end of this expression, leaving just the inner expression.

Template interpolation syntax is still used to construct strings from
expressions when the template includes multiple interpolation sequences or a
mixture of literal strings and interpolations. This deprecation applies only
to templates that consist entirely of a single interpolation sequence.

(and 5 more similar warnings elsewhere)


Warning: Quoted references are deprecated

  on cljdoc.tf line 42, in resource "aws_route53_zone" "cljdoc_xyz_zone":
  42:   provider = "aws.prod"

In this context, references are expected literally rather than in quotes.
Terraform 0.11 and earlier required quotes, but quoted references are now
deprecated and will be removed in a future version of Terraform. Remove the
quotes surrounding this reference to silence this warning.

(and one more similar warning elsewhere)

Terraform has been successfully initialized!

You may now begin working with Terraform. Try running "terraform plan" to see
any changes that are required for your infrastructure. All Terraform commands
should now work.

If you ever set or change modules or backend configuration for Terraform,
rerun this command to reinitialize your working directory. If you forget, other
commands will detect it and remind you to do so if necessary.
```
Here we see some warnings, so I can guess we are at 0.11.

There is some upgrade support in terraform.
```
terraform 0.12upgrade
```
This made some changes.
If we retry `terraform init` we see no more warnings:
```
Initializing modules...

Initializing the backend...

Initializing provider plugins...

The following providers do not have any version constraints in configuration,
so the latest version was installed.

To prevent automatic upgrades to new major versions that may contain breaking
changes, it is recommended to add version = "..." constraints to the
corresponding provider blocks in configuration, with the constraint strings
suggested below.

* provider.aws: version = "~> 2.70"
* provider.digitalocean: version = "~> 2.29"

Terraform has been successfully initialized!

You may now begin working with Terraform. Try running "terraform plan" to see
any changes that are required for your infrastructure. All Terraform commands
should now work.

If you ever set or change modules or backend configuration for Terraform,
rerun this command to reinitialize your working directory. If you forget, other
commands will detect it and remind you to do so if necessary.
```
## v0.13.7
This time we have failures for `terraform init`

```
terraform init
Initializing modules...

Initializing the backend...

Initializing provider plugins...
- Finding latest version of hashicorp/aws...
- Finding latest version of hashicorp/digitalocean...
- Installing hashicorp/aws v5.11.0...
- Installed hashicorp/aws v5.11.0 (signed by HashiCorp)

Warning: Quoted references are deprecated

  on public_bucket/bucket.tf line 4, in data "aws_iam_policy_document" "public_read_bucket_policy":
   4:   provider = "aws.prod"

In this context, references are expected literally rather than in quotes.
Terraform 0.11 and earlier required quotes, but quoted references are now
deprecated and will be removed in a future version of Terraform. Remove the
quotes surrounding this reference to silence this warning.

(and 13 more similar warnings elsewhere)


Warning: Interpolation-only expressions are deprecated

  on public_bucket/bucket.tf line 24, in resource "aws_s3_bucket" "public_bucket":
  24:   bucket   = "${var.bucket_name}"

Terraform 0.11 and earlier required all non-constant expressions to be
provided via interpolation syntax, but this pattern is now deprecated. To
silence this warning, remove the "${ sequence from the start and the }"
sequence from the end of this expression, leaving just the inner expression.

Template interpolation syntax is still used to construct strings from
expressions when the template includes multiple interpolation sequences or a
mixture of literal strings and interpolations. This deprecation applies only
to templates that consist entirely of a single interpolation sequence.

(and 26 more similar warnings elsewhere)


Error: Failed to install providers

Could not find required providers, but found possible alternatives:

  hashicorp/digitalocean -> digitalocean/digitalocean

If these suggestions look correct, upgrade your configuration with the
following commands:
    terraform 0.13upgrade .
    terraform 0.13upgrade server_instance
```

But again, we do have some upgrade support.
```
terraform 0.13upgrade
```
This was not as helpful, it made a change but init still failed.

I addressed warnings manually.

It then suggested I run `terraform 0.13upgrade server_instance`, this created new config.

And then `terraform init` was successful.

## v0.14.11

`terraform init` succeeded but created a new lock file which it suggested I check in, so I did.

## Later versions

The last 0.x version was 0.15.5, `terraform init` worked for it.
After that I jumped to the current terrform v1.5.5, `terraform init` for it passed as well.

```
Initializing the backend...
Initializing modules...

Initializing provider plugins...
- Reusing previous version of hashicorp/aws from the dependency lock file
- Reusing previous version of digitalocean/digitalocean from the dependency lock file
- Using previously-installed hashicorp/aws v5.11.0
- Using previously-installed digitalocean/digitalocean v2.29.0

Terraform has been successfully initialized!

You may now begin working with Terraform. Try running "terraform plan" to see
any changes that are required for your infrastructure. All Terraform commands
should now work.

If you ever set or change modules or backend configuration for Terraform,
rerun this command to reinitialize your working directory. If you forget, other
commands will detect it and remind you to do so if necessary.
```

**Comment by lread** [comment]
Closed by #802
